keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+In Kubernetes components perform calculations on the requested resources of a pod to determine whether a pod fits within the limits of a node or namespace resource quota.  The scheduler is responsible for placing a pod onto a node, the kubelet is responsible for consistently enforcing only pods that fit within its allowed requests are admitted, and the quota subsystem tracks and rejects creations or updates that exceed fixed limits by performing consistent calculations.
+
+The reservation of a pod for a given resource is
+
+    max(max_over_init_containers(resource), sum_over_containers(resource)) + resource_overhead
+
+The reservation of a pod is the union of all resource calculations of this form. Resources are considered individually when making decisions about admission or rejection from nodes or quota, and any one resource being exceeded will prevent execution or admission.
+
+For instance, the pod:
+
+```
+metadata:
+  name: nginx
+spec:
+  initContainers:
+  - name: copy-files
+    ...
+    resources:
+      requests:
+        cpu: 100m
+  - name: generate-config
+    ...
+    resources:
+      requests:
+        cpu: 300m
+  containers:
+  - name: proxy
+    ...
+    resources:
+      requests:
+        cpu: 50m
+  - name: sidecar-logging
+    ...
+    resources:
+      requests:
+        cpu: 100m
+```
+
+will be handled by the system in the following ways:
+
+* Is considered to consume `300m` cores of CPU by the scheduler (the node must be able to run the largest init container by itself, which is larger than the sum of the proxy and sidecar containers at runtime)
+* The kubelet will not admit the pod if `(the node's allocatable CPU) - (the sum of all running pods)` is less than `300m`
+* The kubelet will create a cgroup for the pod that expects to get roughly `300m` cores, but the container cgroup created for `copy-files` will request `100m` while that init container is running
+* The quota subsystem will block this pod from being created if there is less than `300m` available.
+
+Once a pod is created, it passes through four high level lifecycle phases as seen by the total Kubernetes system. The first phase is `Pending` - the time before the pod is scheduled to a node. The second phase is `Initializing` - the time between when the pod is scheduled on a node and when all init containers have completed successfully and the `Initalized` condition on the pod is set to `true`. A pod without init containers may only remain in this phase very briefly. The third phase is `Running`, when all init containers have completed, and continues until the pod is deleted or reaches a terminal state of success or failure.  The final state is `Completed` which means the pod has no running containers, all resources have been released or cleaned up, and the pod will never again consume those resources.
```

Comment:
I think this will need some deeper thought. As it stands this sounds like we will have a lot of very short lived series, which we should prevent.

Reply:
Hrm.  So short lived pods already have this problem for conditions, so this should be strictly less than almost every other ksm style metric that exposes conditions (which change more frequently than this).  Maybe we should remove other short lived series in order to do this, if short lived series budget is a concern?